### PR TITLE
[RLlib] Issue 21991: Fix `SampleBatch` slicing for `SampleBatch.INFOS` in RNN cases

### LINKED ISSUE #21991: [RLlib][Bug] Wrong number of samples in `batch[SampleBatch.INFOS]` in RNN cases
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I'm trying to train my custom models by adding additional custom loss with `info`. This is a common case for training RL policies with auxiliary tasks and online imitation with behavior cloning.

I add some auxiliary information in my environment's `step` method. For example, the field `info` contains the component of the reward function.

```python
class MyEnv(gym.Env):
    def step(self, action):
        # forward dynamic
        reward_1 = reward_func_1()
        reward_2 = reward_func_2()
        reward = reward_1 + reward_2

        info = {'reward_1': reward_1, 'reward_2': reward_2}

        return observation, reward, done, info
```

My custom model with custom loss:

```python
class CustomModel(TorchRNN):
    def custom_loss(self, policy_loss, loss_inputs):

        seq_lens = loss_inputs[SampleBatch.SEQ_LENS]
        infos = loss_inputs[SampleBatch.INFOS]

        B = len(loss_inputs[SampleBatch.SEQ_LENS])
        num_padded_samples = len(loss_inputs[SampleBatch.OBS])

        max_seq_len = num_padded_samples // B
        mask = sequence_mask(seq_lens, maxlen=max_seq_len)

        reward_1 = ... # extract from `loss_inputs[SampleBatch.INFOS]`
        reward_2 = ... # extract from `loss_inputs[SampleBatch.INFOS]`

        reward_1 = ... # padding and add time dimension
        reward_2 = ... # padding and add time dimension

        reward_1_predicted = self.reward_1_prediction_branch(self._features)
        reward_2_predicted = self.reward_2_prediction_branch(self._features)

        custom_loss = ... # reward prediction loss using items from `info`

        return policy_loss + custom_loss
```

In RNN cases, the sample batch will be padded and stacked into `torch.Tensor` before feeding into method `model.forward` and `model.custom_loss`, except `INFOS` and `SEQ_LENS`:

https://github.com/ray-project/ray/blob/b73a007ccde13d809b20252478f086b4ad997e72/rllib/policy/rnn_sequencing.py#L109-L120

For example, if I get a batch with batch size 16 and max sequence length 25, then:

```python
# after padding
batch[SampleBatch.OBS]       # a torch tensor (dtype: float,  shape: (16, 25, *))
batch[SampleBatch.SEQ_LENS]  # a torch tensor (dtype: int,    shape: (16,))
batch[SampleBatch.INFOS]     # a numpy array  (dtype: object, shape: (num_info,))
```

Although for compatibility reasons, RLlib cannot assume the content in `info`. I can make my own batch from `info` by manual padding and reshaping. In sample batch, it should have:

```python
num_infos == seq_lens.sum()
```

### Expected

Ideally:

```python
len(batch[SampleBatch.INFOS]) == batch[SampleBatch.SEQ_LENS].sum().item()
```

I don't really sure about the details of how RLlib construct sample batch for RNNs. Both `rollout_fregment_length` and `max_seq_len` in `config` will affect the input batch. So I think the following is also acceptable:

```python
len(batch[SampleBatch.INFOS]) >= batch[SampleBatch.SEQ_LENS].sum().item()
```

### What Happened

Sometimes (not always), I got:

```python
len(batch[SampleBatch.INFOS]) < batch[SampleBatch.SEQ_LENS].sum().item()
```

This usually happens when `num_vaild_samples != B * T`, i.e., the episode is complete and the batch is padded.

image

I can reproduce this issue with either `"batch_mode": "truncate_episodes"` or `"batch_mode": "complete_episodes"`.

### Versions / Dependencies

Ray nightly

### Reproduction script

```python
import argparse
import os
import sys
from warnings import warn

import ray
from ray import tune
from ray.rllib.examples.models.rnn_model import TorchRNNModel
from ray.rllib.models import ModelCatalog
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.utils.torch_utils import sequence_mask


DEBUG = (getattr(sys, 'gettrace', lambda: None)() is not None)


class TorchRNNModelWithInfos(TorchRNNModel):
    def custom_loss(self, policy_loss, loss_inputs):

        seq_lens = loss_inputs[SampleBatch.SEQ_LENS]
        infos = loss_inputs[SampleBatch.INFOS]

        B = len(loss_inputs[SampleBatch.SEQ_LENS])
        num_padded_samples = len(loss_inputs[SampleBatch.OBS])

        max_seq_len = num_padded_samples // B
        mask = sequence_mask(seq_lens, maxlen=max_seq_len)

        num_valid_samples = mask.sum().item()
        num_infos = len(infos)

        assert len(loss_inputs) == num_valid_samples
        assert seq_lens.sum().item() == num_valid_samples

        if num_infos < num_valid_samples:
            warn(f"Insufficient number of `info` "
                 f"(expected {num_valid_samples}, got {len(infos)}).")
        elif num_infos > num_valid_samples:
            warn(f"Incorrect number of `info` "
                 f"(expected {num_valid_samples}, got {len(infos)}).")

        # assert num_infos >= num_valid_samples

        return policy_loss


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--run", type=str, default="PPO", help="The RLlib-registered algorithm to use."
    )
    parser.add_argument("--env", type=str, default="CartPole-v0")
    parser.add_argument("--num-cpus", type=int, default=0)
    parser.add_argument(
        "--stop-iters", type=int, default=100, help="Number of iterations to train."
    )
    parser.add_argument(
        "--stop-timesteps", type=int, default=100000, help="Number of timesteps to train."
    )

    args = parser.parse_args()

    ray.init(num_cpus=args.num_cpus or None, local_mode=DEBUG)

    ModelCatalog.register_custom_model(
        "rnn", TorchRNNModelWithInfos
    )

    config = {
        "env": args.env,
        "framework": "torch",
        "gamma": 0.9,
        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        "_fake_gpus": not DEBUG,
        "num_workers": 0,
        "num_envs_per_worker": 20,
        "entropy_coeff": 0.001,
        "num_sgd_iter": 5,
        "vf_loss_coeff": 1E-5,
        "model": {
            "custom_model": "rnn",
            "max_seq_len": 20,
            "custom_model_config": {
                "cell_size": 32,
            },
        },

        # "batch_mode": "complete_episodes",
        "batch_mode": "truncate_episodes",
    }

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
    }

    results = tune.run(args.run, config=config, stop=stop, verbose=1)

    ray.shutdown()


if __name__ == "__main__":
    main()
```

I got:

```
(PPOTrainer pid=314050) 2022-01-31 16:48:27,935 WARNING util.py:60 -- Install gputil for GPU system monitoring.
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 130, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 122, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 132, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 125, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 119, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 119, got 100).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 137, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 139, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "

(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 112, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 141, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 136, got 40).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 131, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 117, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 124, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "

(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 131, got 100).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 134, got 80).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 135, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
(PPOTrainer pid=314050) custom_rnn_model.py:36: UserWarning: Insufficient number of `info` (expected 113, got 0).
(PPOTrainer pid=314050)   warn(f"Insufficient number of `info` "
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by XuehaiPan**
## Why are these changes needed?

1. In RNN cases, the sample batch will be padded and stacked into `torch.Tensor` before feeding into method `model.forward` and `model.custom_loss`, except `INFOS` and `SEQ_LENS`:

    https://github.com/ray-project/ray/blob/b73a007ccde13d809b20252478f086b4ad997e72/rllib/policy/rnn_sequencing.py#L109-L120

    The keys `SampleBatch.INFOS`, `SampleBatch.SEQ_LENS`, and `"state_*"` will not be padded.

    Therefore key `SampleBatch.INFOS` needs to be specially treated as well during batch slicing (`batch[start:end]`).

    https://github.com/ray-project/ray/blob/9c95b9a5fae5882b6b89cd80cbc2a55b4b50c405/rllib/policy/sample_batch.py#L924-L930

2. In `SampleBatch._slice_map`:

    https://github.com/ray-project/ray/blob/9c95b9a5fae5882b6b89cd80cbc2a55b4b50c405/rllib/policy/sample_batch.py#L906-L916

    The value for key `SampleBatch.SEQ_LENS` is a tensor. And variable `l` is an integer tensor instead of a Python `int`. The instance `sum_` can be a tensor and will be reused. Because in-place operator `+=` is used.

    Screenshot debugger

    Screenshot watch

3. See https://github.com/ray-project/ray/pull/22050#discussion_r803769472

## Related issue number

Fixes #21991

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [X] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/sample_batch.py [hunk 1]
```diff
@@ -906,26 +906,30 @@ def _slice(self, slice_: slice) -> "SampleBatch":
             # Build our slice-map, if not done already.
             if not self._slice_map:
                 sum_ = 0
-                for i, l in enumerate(self[SampleBatch.SEQ_LENS]):
-                    for _ in range(l):
-                        self._slice_map.append((i, sum_))
-                    sum_ += l
```
Comment (XuehaiPan) [inline rllib/policy/sample_batch.py:912]: Python's built-in type `int` is immutable, it's okay to use `x += y`. Statement `x += y` will create a new `int` instance.

```python
>>> x = 1000

>>> x
1000
>>> id(x)
140102686402576

>>> x += 1
>>> x
1001
>>> id(x)
140102686403152
```

Here `l` and `sum_` are tensors, `sum_ += l` will not create a new instance. `slice_map.append((i, sum_))` is adding the reference to the same instance `sum_` in memory.

```python
>>> import torch
>>> sum_ = 0
>>> l = torch.ones((), dtype=torch.int32)

>>> sum_
0
>>> id(sum_)
94230645960128

>>> sum_ += l
>>> sum_
tensor(1, dtype=torch.int32)
>>> id(sum_)
140100675095104

>>> sum_ += l
>>> sum_
tensor(2, dtype=torch.int32)
>>> id(sum_)
140100675095104
```

File: rllib/policy/sample_batch.py [hunk 2]
```diff
@@ -906,26 +906,30 @@ def _slice(self, slice_: slice) -> "SampleBatch":
...
                 if path[0] != SampleBatch.SEQ_LENS and not path[0].startswith(
                     "state_in_"
                 ):
-                    return value[start:stop]
+                    if path[0] != SampleBatch.INFOS:
+                        return value[start_padded:stop_padded]
+                    else:
+                        return value[start_unpadded:stop_unpadded]
                 else:
                     return value[start_seq_len:stop_seq_len]
```
Comment (XuehaiPan) [inline rllib/policy/sample_batch.py:934]: Special case for `SampleBatch.INFOS` (unpadded).

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 3]
```diff
@@ -730,6 +730,7 @@ def get_inference_input_dict(self, policy_id: PolicyID) -> Dict[str, TensorType]
                 if data_col
                 in [
                     SampleBatch.OBS,
+                    SampleBatch.INFOS,
```
Comment (XuehaiPan) [inline rllib/evaluation/collectors/simple_list_collector.py:733]: In method `get_inference_input_dict()`, the key `INFOS` should be shift togather with `ODS` / `T`.

```python
# initial step
obs = env.reset()
info = {}
t = -1
state = policy.model.get_initial_state()

results = policy.compute_single_action(obs,        # obs  at `t = -1`
                                       info=info,  # info at `t = -1`
                                       state=state)
action, state, *_ = results

# interaction (t = 0)
obs, reward, done, info = env.step(action)
t = 0
results = policy.compute_single_action(obs,        # obs  at `t = 0`
                                       info=info,  # info at `t = 0`
                                       state=state)
action, state, *_ = results

# interaction (t = 1)
obs, reward, done, info = env.step(action)
t = 1
...
```

**Comment by XuehaiPan** [comment]
Hi, @sven1977, could you have a look at this PR? Thanks!

File: rllib/policy/rnn_sequencing.py [hunk 4]
```diff
@@ -115,7 +115,7 @@ def pad_batch_to_sequences_of_same_size(
         elif (
             not feature_keys
             and not k.startswith("state_out_")
-            and k not in ["infos", SampleBatch.SEQ_LENS]
+            and k not in [SampleBatch.INFOS, SampleBatch.SEQ_LENS]
```
Comment (sven1977) [inline rllib/policy/rnn_sequencing.py:118]: Nice!

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:964]: Nice! Thanks for this fix.

**Review by sven1977 — APPROVED** [review]
Thanks for this PR @XuehaiPan and the detailed explanations! :)
